### PR TITLE
Refine AbstractModule methods

### PR DESCRIPTION
**Pull request by yiheng**
## What changes were proposed in this pull request?
This PR is code refactor, include:
1. make getParamter method as final and private[bigdl]. That method should only used in BigDL optimization. If user accident call it outside, the optimization result maybe incorrect
2. Remove updateParameters method. It's only used for debug, should not be a public API. User should use optimizer instead of updateParameter method to change the model weights
3. ZeroGrad should be final, remove the redundant implementations
4. add final to necessary methods
5. reorder the methods, put public method first

## How was this patch tested?
unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/966

### CODE REVIEW

**Comment by zhichao-li** [comment]
One idea is we can provide a common method for `getGrad`, then  `zeroGrad` doent's need to be override in every place as it just zero grads....

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -340,16 +340,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
-   * to use optim method
+   * This function returns two tensors. One for the flattened trainable parameters flatParameters
+   * and another for the gradients of the energy wrt to the trainable parameters flatGradParameters.
+   *
+   * Custom modules should not override this function. They should instead override parameters(...)
+   * which is, in turn, called by the present function.
+   *
+   * This function will go over all the weights and gradWeights and make them view into a single
+   * tensor (one for weights and one for gradWeights). Since the storage of every weight and
+   * gradWeight is changed.
```
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: Feels like this sentence is not finished.

**Comment by yiheng** [comment]
@zhichao-li This PR is removing the unnecessary zeroGrad methods. ZeroGrad is using parameter() method to get the grads. We needn't an extra getGrad methods

**Comment by yiheng** [comment]
jenkins test pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 2]
```diff
@@ -849,30 +777,206 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+  final def evaluate(
+    dataSet: LocalDataSet[MiniBatch[T]],
+    vMethods: Array[ValidationMethod[T]]
+  ): Array[(ValidationResult, ValidationMethod[T])] = {
     Validator(this, dataSet).test(vMethods)
   }
 
-  def quantize(): Module[T] = {
+  /**
+   * Quantize this module, whcih reduces the precison of the parameter. Get a higher speed with a
```
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: which and precision are spelled wrong.

**Comment by yangw1234** [comment]
LGTM

**Comment by yiheng** [comment]
jenkins pass.merge

**Pull request merged**